### ISSUE TITLE
Documents are not clear on testing mechanisms

### ISSUE THREAD

**Issue by gatos-cc** [elastic/apm-agent-python#925]
**Describe the bug**: ...

CONTRIBUTING.md contains the following:

> To run the full test suite, you need to install several databases (Elasticsearch, PostgreSQL, MySQL, Cassandra, Redis). This can be quite a hassle, so we recommend to use our dockerized test setup. See [Running tests](https://www.elastic.co/guide/en/apm/agent/python/master/run-tests-locally.html) for detailed instructions.

README.md also points to [Running tests](https://www.elastic.co/guide/en/apm/agent/python/master/run-tests-locally.html)

Running tests contains the following:

> To run tests locally you can make use of the docker images also used when running the whole test suite with Jenkins. Running the full test suite first does some linting and then runs the actual tests with different versions of Python and different web frameworks. For a full overview of the test matrix and supported versions have a look at Jenkins Configuration.

That document does not however detail *how* to build/start the docker images.  It includes instructions on running `./tests/scripts/docker/run_tests.sh`, but as near as I can tell this is all run entirely on a host machine, not using a docker container.  It also requires Python and several pip installs locally.

I would expect something in one of the docs of the form:
```
docker compose build...
docker build -t ...  && docker run ...
etc
```

Otherwise locally a clearer picture of exactly what resources I need to start up and how to configure them instead of simply "This can be quite a hassle".

Also, CONTRIBUTING.md contains the following:

> However, for running local unit tests, you can install the relevant
[requirements files](https://github.com/elastic/apm-agent-python/tree/master/tests/requirements)
and then run `py.test` from the project root.

The root of the project does not contain a `py.test` file, nor does the test directory or the test/scripts directory.

Basically I'm trying to put together a pull request, and I'm really not sure how to build/test this repo.

**Comment by basepi** [comment]
Hey @gatos-cc! I guess I need to clarify that last note for running `py.test` -- that's a binary script which is installed with `pytest`, one of the requirements in those requirements files.

Honestly we don't expect anyone to run the integration tests before submitting a PR. If you can run the unit tests with pytest, that's plenty. I may do a bit of a restructure on that doc to make that clear.

**Comment by gatos-cc** [comment]
Ah, that's fair.  I'm not normally a Python developer, so some of these things are probably obvious that I'm missing.  Just the nature of this project you're going to have devops type folks coming in that might be in a similar position.

I'll start with installing the requirements file and see where I get - thanks @basepi

**Comment by basepi** [comment]
Sounds great! Please keep me posted if you have any other questions.

Another note: `./tests/scripts/docker/run_tests.sh` actually [does all the `docker compose` stuff](https://github.com/elastic/apm-agent-python/blob/fd32c6c07690f8b5b35a2819cdf18a504d84ba9c/tests/scripts/docker/run_tests.sh#L44-L63) for you. So if you do want to go above and beyond, that script should work -- if it doesn't, we'd like to fix it!

I opened #926 with some clarifications. If you have a moment to review, I'd love to know if those clarifications would have been sufficient for you.

**Comment by gatos-cc** [comment]
Yep - #926 looks good to me.  There are a lot of requirements files in that dir, but it looks like they all pull in reqs-base.txt, so most devs will be fine just picking whatever env they're running and be fine.

**Comment by basepi** [comment]
That's the idea! Excellent, thanks for the follow-up.

**Comment by gatos-cc** [comment]
Hey @basepi - sorry to tack on to a closed issue, but just want to make sure this isn't user error before opening a new issue.  Trying to run the unit tests, I'm on a CentOS docker image, running the following:
```
git clone https://github.com/elastic/apm-agent-python.git
cd apm-agent-python
pip install -r tests/requirements/reqs-flask-1.1.txt
make update-json-schema
py.test
```

The output reports 28 skipped and 11 errors:
```
# py.test
======================================================================================== test session starts ========================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.9, py-1.8.1, pluggy-0.13.1
benchmark: 3.2.2 (defaults: timer=time.time disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /opt/test/tmp/apm-agent-python, inifile: setup.cfg
plugins: bdd-3.2.1, benchmark-3.2.2, mock-2.0.0, django-3.8.0, rerunfailures-8.0, cov-2.8.1, localserver-0.5.0
collected 504 items / 11 errors / 28 skipped / 465 selected

============================================================================================== ERRORS ===============================================================================================
_______________________________________________________________________ ERROR collecting tests/client/py3_exception_tests.py ________________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/client/py3_exception_tests.py", line 51
E       raise ValueError("bla") from zde
E                                  ^
E   SyntaxError: invalid syntax
____________________________________________________________________ ERROR collecting tests/contrib/asyncio/aiohttp_web_tests.py ____________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/contrib/asyncio/aiohttp_web_tests.py", line 49
E       async def hello(request):
E               ^
E   SyntaxError: invalid syntax
_____________________________________________________________________ ERROR collecting tests/contrib/asyncio/starlette_tests.py _____________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/contrib/asyncio/starlette_tests.py", line 53
E       async def hi(request):
E           ^
E   SyntaxError: invalid syntax
__________________________________________________________________ ERROR collecting tests/contrib/asyncio/tornado/tornado_tests.py __________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/contrib/asyncio/tornado/tornado_tests.py", line 109
E       async def test_get(app, base_url, http_client):
E           ^
E   SyntaxError: invalid syntax
___________________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/aiohttp_client_tests.py ____________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/aiohttp_client_tests.py", line 43
E       async def test_http_get(instrument, event_loop, elasticapm_client, waiting_httpserver, use_yarl):
E           ^
E   SyntaxError: invalid syntax
________________________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/aiopg_tests.py ________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/aiopg_tests.py", line 66
E       async def cursor(request):
E           ^
E   SyntaxError: invalid syntax
_____________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/async_elasticsearch_client_tests.py ______________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/async_elasticsearch_client_tests.py", line 52
E       async def async_elasticsearch(request):
E           ^
E   SyntaxError: invalid syntax
_______________________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/asyncpg_tests.py _______________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/asyncpg_tests.py", line 57
E       async def connection(request):
E           ^
E   SyntaxError: invalid syntax
________________________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/base_tests.py _________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/base_tests.py", line 41
E       async def test_async_capture_span(instrument, elasticapm_client):
E               ^
E   SyntaxError: invalid syntax
_____________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/elasticsearch_async_client_tests.py ______________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/elasticsearch_async_client_tests.py", line 52
E       async def elasticsearch_async(request):
E           ^
E   SyntaxError: invalid syntax
________________________________________________________________ ERROR collecting tests/instrumentation/asyncio_tests/httpx_tests.py ________________________________________________________________
/usr/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib/python2.7/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
E     File "/opt/test/tmp/apm-agent-python/tests/instrumentation/asyncio_tests/httpx_tests.py", line 47
E       async def test_httpx_instrumentation(instrument, elasticapm_client, waiting_httpserver):
E               ^
E   SyntaxError: invalid syntax
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 11 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================================================== 28 skipped, 11 error in 5.19 seconds ================================================================================
```

So... Is this user error, or are there legitimate issues?

**Comment by basepi** [comment]
These are all issues stemming from tests that are only compatible with python3. That said, I thought we had automatically had pytest exclude these.....but perhaps on python2 we need to add an explicit exclude as a command line arg. Let me do some looking.

**Comment by basepi** [comment]
Ah, I see. Try doing `make test` instead of `py.test` -- that should automatically add [the required flags](https://github.com/elastic/apm-agent-python/blob/master/Makefile#L17) to `py.test`.

I'll update the contributing docs.

**Comment by gatos-cc** [comment]
Commenting out this guy:
```
#@pytest.mark.skipif(platform.system() == "Windows", reason="os.access() doesn't seem to work as we expect on Windows")
#def test_file_is_readable_validator_not_readable(tmpdir):
#    p = tmpdir.join("nonreadable")
#    p.write("")
#    os.chmod(p.strpath, stat.S_IWRITE)
#    validator = FileIsReadableValidator()
#    with pytest.raises(ConfigurationError) as e:
#        validator(p.strpath, "path")
#    assert "is not readable" in e.value.args[0]
```
everything passes

**Comment by gatos-cc** [comment]
and actually, that's running on my host machine (macOS) not the Amazon Linux container

**Comment by basepi** [comment]
Weird. I've never seen problems with that particular test. What platform are you working on?

**Comment by gatos-cc** [comment]
Yeah - seems odd.  My full test:

Empty directory, add:
* Dockerfile (below)
* Fresh checkout of `apm-agent-python` (master)

Run `docker build -t test .`

Dockerfile:
```
FROM amazonlinux:2018.03
# FROM centos:latest

RUN df -h

ENV PYTHONUNBUFFERED=1

RUN yum update -y
RUN yum install -y python2 python2-devel
RUN yum install -y git wget bash less gcc glibc-devel which make

# For CentOS if desired:
# RUN ln -s /usr/bin/python2 /usr/bin/python
RUN which python

RUN echo "------------------- Acquire pip -------------------" && \
    mkdir -pv /usr/src/pip && \
    cd /usr/src/pip && \
    wget https://bootstrap.pypa.io/get-pip.py && \
    cd /usr/src/pip && \
    python get-pip.py && \
    mkdir /opt/test

WORKDIR /opt/test

COPY ./ ./

RUN cd /opt/test/apm-agent-python && \
    pip install -r tests/requirements/reqs-flask-1.1.txt && \
    make update-json-schema && \
    make test

CMD [ "bash" ]
```

Output:
```
...lots of output...
============================= test session starts ==============================
platform linux2 -- Python 2.7.18, pytest-4.6.9, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python
cachedir: .pytest_cache
benchmark: 3.2.2 (defaults: timer=time.time disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /opt/test/apm-agent-python, inifile: setup.cfg
plugins: bdd-3.2.1, benchmark-3.2.2, mock-2.0.0, django-3.8.0, rerunfailures-8.0, cov-2.8.1, localserver-0.5.0
collecting ... collected 504 items / 28 skipped / 476 selected
...
tests/bdd/api_key_tests.py::test_a_configured_secret_token_is_sent_if_no_api_key_is_configured PASSED [100%]

=================================== FAILURES ===================================
_________________ test_file_is_readable_validator_not_readable _________________

tmpdir = local('/tmp/pytest-of-root/pytest-0/test_file_is_readable_validato2')

    @pytest.mark.skipif(platform.system() == "Windows", reason="os.access() doesn't seem to work as we expect on Windows")
    def test_file_is_readable_validator_not_readable(tmpdir):
        p = tmpdir.join("nonreadable")
        p.write("")
        os.chmod(p.strpath, stat.S_IWRITE)
        validator = FileIsReadableValidator()
        with pytest.raises(ConfigurationError) as e:
>           validator(p.strpath, "path")
E           Failed: DID NOT RAISE <class 'elasticapm.conf.ConfigurationError'>

tests/config/tests.py:271: Failed
=============================== warnings summary ===============================
tests/client/client_tests.py::test_service_info[elasticapm_client0]
...
tests/utils/tests.py::test_deprecation
  /opt/test/apm-agent-python/tests/utils/tests.py:55: DeprecationWarning: Call to deprecated function deprecated_function. Use alternative instead
    def deprecated_function():

-- Docs: https://docs.pytest.org/en/latest/warnings.html
======= 1 failed, 498 passed, 33 skipped, 318 warnings in 37.54 seconds ========
make: *** [test] Error 1
```

**Comment by gatos-cc** [comment]
Not critical for me personally - I can just comment out the test since it's not a regression and I don't think I'll be doing anything intersecting that test.

**Comment by basepi** [comment]
Sure. Still annoying that it's failing. I may have to spin up a container and see what's going wrong, see if I can fix it....

**Comment by basepi** [comment]
This was a bit mystifying to me for awhile, but then it dawned on me: you can't lock `root` out of read access to a file. That's why the test is failing, you need to be running the tests as non-root.

**Comment by gatos-cc** [comment]
Ah.  Good catch.  Normally wouldn't run anything as root in a container, but for the tests I was just letting it go with defaults.

Thanks @basepi - I think I'm good now.
